API.md

```diff
@@ -0,0 +1,198 @@
+# API Reference
+
+Require the TTN Client module:
+
+```java
+import org.thethingsnetwork.java.app.lib.Client;
+```
+
+## Class: Client
+
+Creates an instance of the client:
+
+```java
+Client client = new Client(region, appId, accessKey [, connOpts]);
+```
+
+* `region [String]`: The region (e.g. `eu`) or full hostname (e.g. `eu.thethings.network`) of the handler to connect to.
+* `appId [String]`: The ID of the application to connect to (e.g. `hello-world`).
+* `appAccessKey [String]`: An access key for the application, formatted as base64 (e.g. `'2Z+MU0T5xZCaqsD0bPqOhzA6iygGFoi4FAgMFgBfXSo='`).
+* `connOpts [MqttConnectOptions]`: Some custom configuration of the MQTT connection. This parameter is optional. For example to use TLS download [mqtt-ca.pem](https://preview.console.thethingsnetwork.org/mqtt-ca.pem) and trust it following this guide: [Trust self-signed certificates](http://howardism.org/Technical/Java/SelfSignedCerts.html):
+
+## Event: connect
+
+Emitted on successful connection.
+
+```java
+client.onConnected(new Consumer<MqttClient>() {
+    public void accept(MqttClient client) {
+        System.out.println("connected !");
+    }
+});
+```
+
+* `cb.client [MqttClient]`: MQTT connection wrapper. See [MQTT](http://www.eclipse.org/paho/files/javadoc/org/eclipse/paho/client/mqttv3/MqttClient.html).
+
+## Event: error
+
+Emitted when the client cannot connect or when a parsing error occurs.
+
+```java
+client.onError(new Consumer<Throwable>() {
+    public void accept(Throwable error) {
+        System.err.println("error: " + error.getMessage());
+    }
+});
+```
+
+* `cb.error [Throwable]`: Error object. See [MQTT](https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html).
+
+## Event: message
+
+Emitted when TTN forwards a message addressed to your application.
+
+```java
+client.onUplink(new BiConsumer<String, Object>() {
+    public void accept(String devId, Object data) {
+        System.out.println("Message: " + devId + " " + data);
+});
+```
+
+* `cb.devId [String]`: Device ID, e.g.: `my-uno`.
+* `cb.data [Object]`: Message data, e.g.:
+
+  ```json
+  {
+    "port": 1,
+    "counter": 10,
+    "payload_raw": "MQ==",
+    "payload_fields": {
+      "led": true
+    },
+    "metadata": {
+      "time": "2016-09-07T12:50:07.068771281Z",
+      "frequency": 868.1,
+      "modulation": "LORA",
+      "data_rate": "SF7BW125",
+      "coding_rate": "4/5",
+      "gateways": [{
+        "gtw_id": "0000024B08060112",
+        "timestamp": 3955426155,
+        "time": "2016-09-07T12:50:07.053048Z",
+        "channel": 4,
+        "rssi": -109,
+        "snr": 5.8,
+        "rf_chain": 1
+      }]
+    }
+  }
+  ```
+
+### Listen for a specific device
+
+```java
+client.onUplink(new BiConsumer<String, Object>() {
+    public void accept(String devId, Object data) {
+        System.out.println("Message: " + devId + " " + data);
+}, "my-uno");
```

Comment:
Any chance the order can be as NodeJS lib? So first optional devId (and field), then callback?